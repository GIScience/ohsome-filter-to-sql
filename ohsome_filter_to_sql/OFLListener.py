# Generated from OFL.g4 by ANTLR 4.13.2
from antlr4 import *
if "." in __name__:
    from .OFLParser import OFLParser
else:
    from OFLParser import OFLParser

# This class defines a complete listener for a parse tree produced by OFLParser.
class OFLListener(ParseTreeListener):

    # Enter a parse tree produced by OFLParser#root.
    def enterRoot(self, ctx:OFLParser.RootContext):
        pass

    # Exit a parse tree produced by OFLParser#root.
    def exitRoot(self, ctx:OFLParser.RootContext):
        pass


    # Enter a parse tree produced by OFLParser#expression.
    def enterExpression(self, ctx:OFLParser.ExpressionContext):
        pass

    # Exit a parse tree produced by OFLParser#expression.
    def exitExpression(self, ctx:OFLParser.ExpressionContext):
        pass


    # Enter a parse tree produced by OFLParser#tagMatch.
    def enterTagMatch(self, ctx:OFLParser.TagMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#tagMatch.
    def exitTagMatch(self, ctx:OFLParser.TagMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#tagWildcardMatch.
    def enterTagWildcardMatch(self, ctx:OFLParser.TagWildcardMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#tagWildcardMatch.
    def exitTagWildcardMatch(self, ctx:OFLParser.TagWildcardMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#tagListMatch.
    def enterTagListMatch(self, ctx:OFLParser.TagListMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#tagListMatch.
    def exitTagListMatch(self, ctx:OFLParser.TagListMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#tagNotMatch.
    def enterTagNotMatch(self, ctx:OFLParser.TagNotMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#tagNotMatch.
    def exitTagNotMatch(self, ctx:OFLParser.TagNotMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#tagNotWildcardMatch.
    def enterTagNotWildcardMatch(self, ctx:OFLParser.TagNotWildcardMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#tagNotWildcardMatch.
    def exitTagNotWildcardMatch(self, ctx:OFLParser.TagNotWildcardMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#hashtagMatch.
    def enterHashtagMatch(self, ctx:OFLParser.HashtagMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#hashtagMatch.
    def exitHashtagMatch(self, ctx:OFLParser.HashtagMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#hashtagWildcardMatch.
    def enterHashtagWildcardMatch(self, ctx:OFLParser.HashtagWildcardMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#hashtagWildcardMatch.
    def exitHashtagWildcardMatch(self, ctx:OFLParser.HashtagWildcardMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#hashtagListMatch.
    def enterHashtagListMatch(self, ctx:OFLParser.HashtagListMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#hashtagListMatch.
    def exitHashtagListMatch(self, ctx:OFLParser.HashtagListMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#typeMatch.
    def enterTypeMatch(self, ctx:OFLParser.TypeMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#typeMatch.
    def exitTypeMatch(self, ctx:OFLParser.TypeMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#idMatch.
    def enterIdMatch(self, ctx:OFLParser.IdMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#idMatch.
    def exitIdMatch(self, ctx:OFLParser.IdMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#typeIdMatch.
    def enterTypeIdMatch(self, ctx:OFLParser.TypeIdMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#typeIdMatch.
    def exitTypeIdMatch(self, ctx:OFLParser.TypeIdMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#idRangeMatch.
    def enterIdRangeMatch(self, ctx:OFLParser.IdRangeMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#idRangeMatch.
    def exitIdRangeMatch(self, ctx:OFLParser.IdRangeMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#idListMatch.
    def enterIdListMatch(self, ctx:OFLParser.IdListMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#idListMatch.
    def exitIdListMatch(self, ctx:OFLParser.IdListMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#typeIdListMatch.
    def enterTypeIdListMatch(self, ctx:OFLParser.TypeIdListMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#typeIdListMatch.
    def exitTypeIdListMatch(self, ctx:OFLParser.TypeIdListMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#geometryMatch.
    def enterGeometryMatch(self, ctx:OFLParser.GeometryMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#geometryMatch.
    def exitGeometryMatch(self, ctx:OFLParser.GeometryMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#areaRangeMatch.
    def enterAreaRangeMatch(self, ctx:OFLParser.AreaRangeMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#areaRangeMatch.
    def exitAreaRangeMatch(self, ctx:OFLParser.AreaRangeMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#perimeterRangeMatch.
    def enterPerimeterRangeMatch(self, ctx:OFLParser.PerimeterRangeMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#perimeterRangeMatch.
    def exitPerimeterRangeMatch(self, ctx:OFLParser.PerimeterRangeMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#lengthRangeMatch.
    def enterLengthRangeMatch(self, ctx:OFLParser.LengthRangeMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#lengthRangeMatch.
    def exitLengthRangeMatch(self, ctx:OFLParser.LengthRangeMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#geometryVerticesRangeMatch.
    def enterGeometryVerticesRangeMatch(self, ctx:OFLParser.GeometryVerticesRangeMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#geometryVerticesRangeMatch.
    def exitGeometryVerticesRangeMatch(self, ctx:OFLParser.GeometryVerticesRangeMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#geometryOutersMatch.
    def enterGeometryOutersMatch(self, ctx:OFLParser.GeometryOutersMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#geometryOutersMatch.
    def exitGeometryOutersMatch(self, ctx:OFLParser.GeometryOutersMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#geometryOutersRangeMatch.
    def enterGeometryOutersRangeMatch(self, ctx:OFLParser.GeometryOutersRangeMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#geometryOutersRangeMatch.
    def exitGeometryOutersRangeMatch(self, ctx:OFLParser.GeometryOutersRangeMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#geometryInnersMatch.
    def enterGeometryInnersMatch(self, ctx:OFLParser.GeometryInnersMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#geometryInnersMatch.
    def exitGeometryInnersMatch(self, ctx:OFLParser.GeometryInnersMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#geometryInnersRangeMatch.
    def enterGeometryInnersRangeMatch(self, ctx:OFLParser.GeometryInnersRangeMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#geometryInnersRangeMatch.
    def exitGeometryInnersRangeMatch(self, ctx:OFLParser.GeometryInnersRangeMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#changesetMatch.
    def enterChangesetMatch(self, ctx:OFLParser.ChangesetMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#changesetMatch.
    def exitChangesetMatch(self, ctx:OFLParser.ChangesetMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#changesetListMatch.
    def enterChangesetListMatch(self, ctx:OFLParser.ChangesetListMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#changesetListMatch.
    def exitChangesetListMatch(self, ctx:OFLParser.ChangesetListMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#changesetRangeMatch.
    def enterChangesetRangeMatch(self, ctx:OFLParser.ChangesetRangeMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#changesetRangeMatch.
    def exitChangesetRangeMatch(self, ctx:OFLParser.ChangesetRangeMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#changesetCreatedByMatch.
    def enterChangesetCreatedByMatch(self, ctx:OFLParser.ChangesetCreatedByMatchContext):
        pass

    # Exit a parse tree produced by OFLParser#changesetCreatedByMatch.
    def exitChangesetCreatedByMatch(self, ctx:OFLParser.ChangesetCreatedByMatchContext):
        pass


    # Enter a parse tree produced by OFLParser#string.
    def enterString(self, ctx:OFLParser.StringContext):
        pass

    # Exit a parse tree produced by OFLParser#string.
    def exitString(self, ctx:OFLParser.StringContext):
        pass



del OFLParser