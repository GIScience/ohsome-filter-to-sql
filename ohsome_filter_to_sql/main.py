import json
import sys
from collections import deque

from antlr4 import CommonTokenStream, InputStream, ParserRuleContext, ParseTreeWalker
from antlr4.error.ErrorListener import ErrorListener

from ohsome_filter_to_sql.OFLLexer import OFLLexer
from ohsome_filter_to_sql.OFLListener import OFLListener
from ohsome_filter_to_sql.OFLParser import OFLParser


class ParserValueError(ValueError):
    pass


class LexerValueError(ValueError):
    pass


class OFLParserErrorListener(ErrorListener):
    def syntaxError(self, recognizer, offendingSymbol, line, column, msg, e):
        # message is based on antlr4 ConsoleErrorListener
        raise ParserValueError("line " + str(line) + ":" + str(column) + " " + msg)


class OFLLexerErrorListener(ErrorListener):
    def syntaxError(self, recognizer, offendingSymbol, line, column, msg, e):
        raise LexerValueError("line " + str(line) + ":" + str(column) + " " + msg)


class OFLToSql(OFLListener):
    def __init__(self):
        self.stack: deque = deque()

    def exitString(self, ctx: ParserRuleContext):
        if ctx.getChildCount() == 1:
            self.stack.append(unescape(ctx.getChild(0).getText()))
        else:
            result = ""
            for child in ctx.getChildren():
                result += child.getText()
            self.stack.append(result)

    # --- methods are sorted in the same order as rules in OFL.g4
    #
    def exitExpression(self, ctx: ParserRuleContext):
        """Handle expression compositions: (), NOT, AND, OR"""
        if ctx.getChildCount() == 3:
            if ctx.getChild(0).getText() == ("(") and ctx.getChild(2).getText() == ")":
                expression = self.stack.pop()
                self.stack.append("(" + expression + ")")
            else:
                right = self.stack.pop()
                left = self.stack.pop()
                operator = ctx.getChild(1).getText().upper()  # AND, OR
                self.stack.append(f"{left} {operator} {right}")
        elif ctx.getChildCount() == 2:
            operator = ctx.getChild(0).getText().upper()  # NOT
            self.stack.append(f"{operator} {self.stack.pop()}")

    # ---
    #
    def exitHashtagMatch(self, ctx: ParserRuleContext):
        hashtag = self.stack.pop()
        self.stack.append(f"'{hashtag}' = any(changeset_hashtags) ")

    def exitHashtagWildcardMatch(self, ctx: ParserRuleContext):
        # TODO
        pass

    def exitHashtagListMatch(self, ctx: ParserRuleContext):
        values = []
        children = list(child.getText() for child in ctx.getChildren())
        # skip first part denoting "hashtag:(" and last part closing list with ")"
        for child in children[3:-1]:
            # skip comma in list in between brackets
            if child == ",":
                continue
            # remove STRING from stack
            self.stack.pop()
            values.append(child)
        values_as_string = "array['" + "', '".join(values) + "']"
        # anyarray && anyarray â†’ boolean (Do the arrays overlap?)
        self.stack.append(f"{values_as_string} && changeset_hashtags")

    # ---
    #
    def exitTagMatch(self, ctx: ParserRuleContext):
        value = self.stack.pop()
        key = self.stack.pop()

        j = json.dumps({key: value})
        # @> Does the first JSON value contain the second?
        self.stack.append(f"tags @> '{j}'")

    def exitTagValuePatternMatch(self, ctx):
        value = (
            self.stack.pop().replace("%", "\\%").replace("_", "\\_").replace("'", "''")
        )
        key = self.stack.pop().replace("'", "''")

        value_child = ctx.getChild(2)
        if value_child.getChild(0).getText() == "*":
            value = "%" + value

        if value_child.getChild(value_child.getChildCount() - 1).getText() == "*":
            value = value + "%"

        self.stack.append(f"tags ->> '{key}' LIKE '{value}'")

    def exitTagWildcardMatch(self, ctx: ParserRuleContext):
        key = self.stack.pop()
        self.stack.append(f"tags ? '{key}'")

    def exitTagNotMatch(self, ctx: ParserRuleContext):
        value = self.stack.pop()
        key = self.stack.pop()
        j = json.dumps({key: value})
        self.stack.append(f"NOT tags @> '{j}'")

    def exitTagNotWildcardMatch(self, ctx: ParserRuleContext):
        key = self.stack.pop()
        self.stack.append(f"NOT tags ? '{key}'")

    def exitTagListMatch(self, ctx: ParserRuleContext):
        values = []
        children = list(child.getText() for child in ctx.getChildren())
        # skip first part denoting "key in (" as well as last part closing list with ")"
        for child in children[3:-1]:
            # skip comma in list in between brackets
            if child == ",":
                continue
            # remove STRING from stack wich are part of *ListMatch
            self.stack.pop()
            values.append(unescape(child))
        values_as_string = "\"', '\"".join(values)
        key = self.stack.pop()
        self.stack.append(f"tags -> '{key}' IN ('\"{values_as_string}\"')")

    # ---
    #
    def exitTypeMatch(self, ctx: ParserRuleContext):
        type_ = ctx.getChild(2).getText()  # NODE, WAY, RELATION
        self.stack.append(f"osm_type = '{type_}'")

    def exitIdMatch(self, ctx: ParserRuleContext):
        id = ctx.getChild(2).getText()
        self.stack.append(f"osm_id = {id}")

    def exitTypeIdMatch(self, ctx: ParserRuleContext):
        type_, id = ctx.getChild(2).getText().split("/")
        self.stack.append(f"(osm_type = '{type_}' AND osm_id = {id})")

    def exitIdRangeMatch(self, ctx: ParserRuleContext):
        child = ctx.getChild(3).getText()
        lower_bound, upper_bound = child.split("..")
        if lower_bound and upper_bound:
            self.stack.append(f"(osm_id >= {lower_bound} AND osm_id <= {upper_bound})")
        elif lower_bound:
            self.stack.append(f"osm_id >= {lower_bound}")
        elif upper_bound:
            self.stack.append(f"osm_id <= {upper_bound}")

    def exitIdListMatch(self, ctx: ParserRuleContext):
        # differs from TagListMatch insofar that no STRING needs to be popped from stack
        values = []
        children = list(child.getText() for child in ctx.getChildren())
        # skip first part denoting "id:(" as well as last part closing list with ")"
        for child in children[3:-1]:
            # skip comma in list in between brackets
            if child == ",":
                continue
            values.append(child)
        values_as_string = ", ".join(values)
        self.stack.append(f"osm_id IN ({values_as_string})")

    def exitTypeIdListMatch(self, ctx: ParserRuleContext):
        values = []
        children = list(child.getText() for child in ctx.getChildren())
        # skip first part denoting "id:(" as well as last part closing list with ")"
        for child in children[3:-1]:
            # skip comma in list in between brackets
            if child == ",":
                continue
            type_, id = child.split("/")
            values.append(f"(osm_id = {id} AND osm_type = '{type_}')")
        if len(values) > 1:
            values_as_string = "(" + " OR ".join(values) + ")"
        else:
            values_as_string = " OR ".join(values)
        self.stack.append(values_as_string)

    # ---
    #
    def exitGeometryMatch(self, ctx: ParserRuleContext):
        geom_type = ctx.getChild(2).getText()
        match geom_type:
            case "point":
                self.stack.append("(status_geom_type).geom_type = 'Point'")
            case "line":
                self.stack.append("(status_geom_type).geom_type = 'LineString'")
            case "polygon":
                self.stack.append(
                    "((status_geom_type).geom_type = 'Polygon' "
                    + "OR (status_geom_type).geom_type = 'MultiPolygon')"
                )
            case "other":
                self.stack.append("(status_geom_type).geom_type = 'GeometryCollection'")

    def exitAreaRangeMatch(self, ctx: ParserRuleContext):
        child = ctx.getChild(3).getText()
        lower_bound, upper_bound = child.split("..")
        if lower_bound and upper_bound:
            self.stack.append(f"(area >= {lower_bound} AND area <= {upper_bound})")
        elif lower_bound:
            self.stack.append(f"area >= {lower_bound}")
        elif upper_bound:
            self.stack.append(f"area <= {upper_bound}")

    def exitLengthRangeMatch(self, ctx: ParserRuleContext):
        child = ctx.getChild(3).getText()
        lower_bound, upper_bound = child.split("..")
        if lower_bound and upper_bound:
            self.stack.append(f"(length >= {lower_bound} AND length <= {upper_bound})")
        elif lower_bound:
            self.stack.append(f"length >= {lower_bound}")
        elif upper_bound:
            self.stack.append(f"length <= {upper_bound}")

    def exitGeometryVerticesRangeMatch(self, ctx: ParserRuleContext):
        # TODO
        pass

    def exitGeometryOutersMatch(self, ctx: ParserRuleContext):
        # TODO
        pass

    def exitGeometryOutersRangeMatch(self, ctx: ParserRuleContext):
        # TODO
        pass

    def exitGeometryInnersMatch(self, ctx: ParserRuleContext):
        # TODO
        pass

    def exitGeometryInnersRangeMatch(self, ctx: ParserRuleContext):
        # TODO
        pass

    # ---
    #
    def exitChangesetMatch(self, ctx: ParserRuleContext):
        id = ctx.getChild(2).getText()
        self.stack.append(f"changeset_id = {id}")

    def exitChangesetListMatch(self, ctx: ParserRuleContext):
        values = []
        children = list(child.getText() for child in ctx.getChildren())
        # skip first part denoting "id:(" as well as last part closing list with ")"
        for child in children[3:-1]:
            # skip comma in list in between brackets
            if child == ",":
                continue
            values.append(child)
        values_as_string = ", ".join(values)
        self.stack.append(f"changeset_id IN ({values_as_string})")

    def exitChangesetRangeMatch(self, ctx: ParserRuleContext):
        child = ctx.getChild(3).getText()
        lower_bound, upper_bound = child.split("..")
        if lower_bound and upper_bound:
            self.stack.append(
                f"(changeset_id >= {lower_bound} AND changeset_id <= {upper_bound})"
            )
        elif lower_bound:
            self.stack.append(f"changeset_id >= {lower_bound}")
        elif upper_bound:
            self.stack.append(f"changeset_id <= {upper_bound}")

    def exitChangesetCreatedByMatch(self, ctx: ParserRuleContext):
        editor = self.stack.pop()
        j = json.dumps({"created_by": editor})
        self.stack.append(f"changeset_tags @> '{j}'")


def unescape(string: str):
    if string[0] == '"':
        string = string[1:-1]
        # fmt: off
        string = string.replace("\\\"", "\"") # noqa: Q003
        # fmt: on
        string = string.replace("\\\\", "\\")
        string = string.replace("\\\r", "\r")
        string = string.replace("\\\n", "\n")
    return string


def ohsome_filter_to_sql(filter: str) -> str:
    listener = OFLToSql()
    tree = build_tree(filter)
    translation = walk_tree(tree, listener).stack
    return " ".join(translation)


def cli():
    print(ohsome_filter_to_sql(input()))


def build_tree(filter: str) -> ParserRuleContext:
    """Build a antlr4 parse tree.

    https://github.com/antlr/antlr4/blob/master/doc/listeners.md
    """
    input_stream = InputStream(filter)
    lexer = OFLLexer(input_stream)
    lexer.removeErrorListeners()
    lexer.addErrorListener(OFLLexerErrorListener())
    stream = CommonTokenStream(lexer)
    parser = OFLParser(stream)
    parser.removeErrorListeners()
    parser.addErrorListener(OFLParserErrorListener())
    tree = parser.root()
    return tree


def walk_tree(tree: ParserRuleContext, listener: OFLToSql) -> OFLToSql:
    walker = ParseTreeWalker()
    walker.walk(listener, tree)
    return listener


if __name__ == "__main__":
    ohsome_filter_to_sql(sys.argv[1])
