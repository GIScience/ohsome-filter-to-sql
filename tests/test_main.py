from typing import AsyncGenerator

import asyncpg
import pytest
import pytest_asyncio
from asyncpg import Record
from asyncpg.connection import Connection
from pytest_approval import verify

from ohsome_filter_to_sql.main import build_tree, ohsome_filter_to_sql

pytestmark = pytest.mark.asyncio  # mark all tests


async def validate_and_verify(
    db_con: Connection,
    sql_where_clause: str,
    filter: str,
):
    """Validate query and verify results.

    Build SQL query from given WHERE Clause.
    Run SQL query against a real database.
    Verify SQL query results through approval testing.
    """
    sql = "SELECT COUNT(*) FROM contributions WHERE " + sql_where_clause
    results: list[Record] = await db_con.fetch(sql)
    text = "\n\n".join([filter, sql_where_clause, str(results)])
    return verify(text)


@pytest_asyncio.fixture(scope="session")
async def db_con() -> AsyncGenerator[Connection, None]:
    con = await asyncpg.connect(
        user="postgres",
        password="mylocalpassword",  # noqa: S106
        host="localhost",
        port=5432,
    )
    yield con
    await con.close()


@pytest.mark.parametrize(
    "filter",
    (
        "natural=tree",
        "type:node and natural=tree",
        "id:(node/1, way/2) and type:way",
    ),
)
async def test_build_tree(filter):
    tree = build_tree(filter).toStringTree()
    assert verify(tree)


# -- Tests are sorted in the same order as rules in OFL.g4
#


async def test_expression_and_expression(db_con):
    filter = "natural=tree and leaf_type=broadleaved"
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


async def test_expression_or_expression(db_con):
    filter = "natural=tree or leaf_type=broadleaved"
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.parametrize(
    "filter",
    (
        "(natural=tree)",
        '("addr:housenumber"=45 and "addr:street"="Berliner Stra√üe") or name="HeiGIT"',
    ),
)
async def test_expression_in_brakets(db_con, filter):
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.skip("Not implemented yet.")
async def test_hashtag_match(db_con):
    filter = "hashtag:missingmaps"
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.skip("Not implemented yet.")
@pytest.mark.parametrize(
    "filter",
    (
        "hashtag:(missingmaps)",
        "hashtag:(missingmaps, type, other)",
    ),
)
async def test_hashtag_list_match(db_con, filter):
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


# TODO
# add tests for all grammar keywords like number, geometry, id, ...
@pytest.mark.parametrize(
    "filter",
    (
        '"addr:housenumber"="45"',  # tagMatch
        "natural=tree",  #      tagMatch
        '"type"=boundary',  #   tagMatch w/ keyword as key
        '"building:material"="other"',  #      tagMatch w/ keyword as value
        "natural=*",  #         tagWildcardMatch
        "oneway!=yes",  #       tagNotMatch
        "name!=*",  #            tagNotWildcardMatch
        "highway in (residential, living_street)",  # tagListMatch
        '"type" in (boundary, route)',  #             tagListMatch w/keyword as key
        "highway in (residential, other)",  #   tagListMatch w/keyword unquoted as value
        "natural in (water)",  #                tagListMatch w/keyword with single value
        '"*"=*',
    ),
)
async def test_tag_match(db_con, filter):
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.parametrize(
    "filter",
    (
        "type:node",
        "type:way",
        "type:relation",
    ),
)
async def test_type_match(db_con, filter):
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


async def test_id_match(db_con):
    filter = "id:4540889804"
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.parametrize(
    "filter",
    (
        "id:node/4540889804",
        "id:way/1136431018",
        "id:relation/2070281",
    ),
)
async def test_type_id_match(db_con, filter):
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.parametrize(
    "filter",
    (
        "id:(1..9999)",
        "id:(..9999)",
        "id:(1..)",
    ),
)
async def test_id_range_match(db_con, filter):
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.parametrize(
    "filter",
    (
        "id:(4540889804)",
        "id:(1136431018, 4540889804, 2070281)",
    ),
)
async def test_id_list_match(db_con, filter):
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.parametrize(
    "filter",
    (
        "id:(node/4540889804)",
        "id:(node/4540889804, way/1136431018, relation/2070281)",
    ),
)
async def test_type_id_list_match(db_con, filter):
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.parametrize(
    "filter",
    (
        "geometry:point",
        "geometry:line",
        "geometry:polygon",
    ),
)
async def test_geometry_match(db_con, filter):
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.skip("Not implemented yet.")
async def test_geometry_match_other(db_con, filter):
    filter = "geometry:other"
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.parametrize(
    "filter",
    (
        "area:(1.0..1E6)",
        "area:(1.3..)",
        "area:(2.0..1.0)",
        "area:(1e6..)",
        "area:(1..1e6)",
    ),
)
async def test_area_range_match(db_con, filter):
    # TODO: ValueError: not enough values to unpack (expected 2, got 1)
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.parametrize(
    "filter",
    (
        "length:(1.0..99.99)",
        "length:(1E6..)",
        "length:(1e6..)",
        "length:(..10000)",
    ),
)
async def test_length_range_match(db_con, filter):
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.skip("Not implemented yet.")
async def test_changeset_match(db_con):
    filter = "changeset:1"
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.skip("Not implemented yet.")
@pytest.mark.parametrize(
    "filter",
    (
        "changeset:(1)",
        "changeset:(1, 300, 4264)",
    ),
)
async def test_changeset_list_match(db_con, filter):
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.skip("Not implemented yet.")
@pytest.mark.parametrize(
    "filter",
    (
        "changeset:(1..999)",
        "changeset:(1..)",
        "changeset:(..999)",
    ),
)
async def test_changeset_range_match(db_con, filter):
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.skip("Not implemented yet.")
@pytest.mark.parametrize(
    "filter",
    (
        "changeset.created_by:Potlatch",
        'changeset.created_by:"Go Map!! 4.3.0"',
        'changeset.created_by:"bulk_upload.py"',
        'changeset.created_by:"JOSM/1.5 (19253 en_GB)"',
    ),
)
async def test_changeset_created_by_match(db_con, filter):
    sql = ohsome_filter_to_sql(filter)
    assert await validate_and_verify(db_con, sql, filter)


@pytest.mark.parametrize(
    "filter",
    (
        "id:(1 ..2",  # missing closing bracket
        "id:(1, 2",  # missing closing bracket
        "id:(1, 2,)",  # missing closing bracket
        "area:(-1..)",
        "area:(1.0..-200)",
        "length:(..-200)",
        "*=*",
    ),
)
async def test_invalid_filters(filter):
    with pytest.raises(ValueError) as e:  # noqa: F841
        ohsome_filter_to_sql(filter)
