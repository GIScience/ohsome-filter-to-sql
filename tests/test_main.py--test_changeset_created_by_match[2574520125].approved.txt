changeset_tags @> '{"created_by": "bulk_upload.py"}'
